fastcrypto-data/src/merkle_tree.rs

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains an implementation of a Merkle Tree data structure (Merkle, R.C. (1988): A Digital Signature
+//! Based on a Conventional Encryption Function) which is a data structure that allows an arbitrary
+//! number of elements of a given type `T` to be added as leaves to the tree and we can then construct
+//! proofs logarithmic in the number of leaves that a certain leaf has a given value. Such proofs can
+//! be verified by a small verifier which only needs to know the root of the tree.
+//!
+//! # Example
+//! ```rust
+//! # use fastcrypto::hash::Sha256;
+//! # use fastcrypto_data::merkle_tree::*;
+//! let elements = [[1u8], [2u8], [3u8]];
+//! let mut tree = MerkleTree::<32, Sha256, [u8; 1]>::new();
+//! tree.insert_all(elements.iter());
+//!
+//! let index = 1;
+//! let proof = tree.prove(index);
+//!
+//! let verifier = tree.create_verifier().unwrap();
+//! assert!(verifier.verify(index, &elements[index], &proof));
+//! ```
+
+use std::borrow::Borrow;
```

Comment:
What is the use case for this data structure? there are many variants of merkle trees, e.g., key-value, with deletions, etc.
Also, what is the expected number of objects, and the expected insertion/deletion pattern?